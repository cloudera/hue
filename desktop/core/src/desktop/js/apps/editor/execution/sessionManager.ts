// Licensed to Cloudera, Inc. under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  Cloudera, Inc. licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import {
  AuthRequest,
  closeSession,
  createSession,
  Session,
  SessionProperty
} from 'apps/editor/execution/api';
import huePubSub from 'utils/huePubSub';

class SessionManager {
  knownSessionPromises: { [key: string]: Promise<Session> } = {};

  /**
   * Gets an existing session or creates a new one if there is no session.
   */
  async getSession(options: { type: string; properties?: SessionProperty[] }): Promise<Session> {
    if (!this.knownSessionPromises[options.type]) {
      this.knownSessionPromises[options.type] = this.createDetachedSession(options);

      // Sessions that fail
      this.knownSessionPromises[options.type].catch(() => {
        delete this.knownSessionPromises[options.type];
      });
    }
    return this.knownSessionPromises[options.type];
  }

  /**
   * Creates a new detached session
   *
   * @param {Object} options
   * @param {String} options.type
   * @param {boolean} [options.preventAuthModal] - Default false
   * @param {Array<SessionProperty>} [options.properties] - Default []
   *
   * @return {Promise<Session>}
   */
  async createDetachedSession(options: {
    type: string;
    properties?: SessionProperty[];
    preventAuthModal?: boolean;
  }): Promise<Session> {
    return new Promise(async (resolve, reject) => {
      const sessionToCreate = {
        type: options.type,
        properties: options.properties || []
      };

      const sessionOrAuth = await createSession(sessionToCreate);
      if ('auth' in sessionOrAuth && sessionOrAuth.auth) {
        const auth = sessionOrAuth as AuthRequest;
        if (!options.preventAuthModal) {
          huePubSub.publish('show.session.auth.modal', {
            message: auth.message,
            session: sessionToCreate,
            resolve: resolve,
            reject: reject
          });
        } else {
          reject(auth);
        }
      } else {
        resolve(sessionOrAuth as Session);
      }
    });
  }

  updateSession(session: Session) {
    this.knownSessionPromises[session.type] = Promise.resolve(session);
  }

  async getAllSessions(): Promise<Session[]> {
    const promises = Object.keys(this.knownSessionPromises).map(
      key => this.knownSessionPromises[key]
    );
    return Promise.all(promises);
  }

  async restartSession(session: Session): Promise<Session> {
    await this.closeSession(session);
    return this.getSession(session);
  }

  hasSession(type: string): boolean {
    return !!this.knownSessionPromises[type];
  }

  async closeSession(session: Session): Promise<void> {
    if (this.hasSession(session.type)) {
      delete this.knownSessionPromises[session.type];
    }
    await closeSession({ session: session, silenceErrors: true });
  }
}

const instance = new SessionManager();

export default instance;
