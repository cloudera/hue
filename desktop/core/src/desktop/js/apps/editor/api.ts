// Licensed to Cloudera, Inc. under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  Cloudera, Inc. licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { format } from '@gethue/sql-formatter';
import { ExecutionStatus } from './execution/sqlExecutable';
import { CancellablePromise } from 'api/cancellablePromise';
import { get } from 'api/utils';

const HISTORY_API_URL = '/api/editor/get_history';

export const formatSql = async (options: {
  statements: string;
  silenceErrors?: boolean;
}): Promise<string> => {
  try {
    return format(options.statements, {
      uppercase: true,
      linesBetweenQueries: 2,
      indentQuerySeparator: true
    });
  } catch (err) {
    if (!options.silenceErrors) {
      throw err;
    }
  }
  return options.statements;
};

export interface FetchHistoryOptions {
  type: string;
  page?: number;
  limit?: number;
  docFilter?: string;
  isNotificationManager?: boolean;
}

export interface FetchHistoryResponse {
  count: number;
  history: {
    absoluteUrl: string;
    data: {
      lastExecuted: number;
      parentSavedQueryUuid: string;
      statement: string;
      status: ExecutionStatus;
    };
    id: number;
    name: string;
    type: string;
    uuid: string;
  }[];
  message: string;
  status: number;
}

export const fetchHistory = (
  options: FetchHistoryOptions
): CancellablePromise<FetchHistoryResponse> => {
  return get<FetchHistoryResponse>(HISTORY_API_URL, {
    doc_type: options.type,
    limit: options.limit || 50,
    page: options.page || 1,
    doc_text: options.docFilter,
    is_notification_manager: options.isNotificationManager
  });
};
