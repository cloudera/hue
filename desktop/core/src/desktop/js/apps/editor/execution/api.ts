// Licensed to Cloudera, Inc. under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  Cloudera, Inc. licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import KnockoutObservable from '@types/knockout';

import {
  EXECUTABLE_RESULT_UPDATED_TOPIC,
  EXECUTABLE_TRANSITIONED_TOPIC,
  ExecutableResultUpdatedEvent,
  ExecutableTransitionedEvent
} from './events';
import Executor from './executor';
import { DefaultApiResponse, extractErrorMessage, post, successResponseIsError } from 'api/utils';
import SqlExecutable, {
  ExecutableContext,
  ExecutionStatus
} from 'apps/editor/execution/sqlExecutable';
import { ResultRow, ResultType } from 'apps/editor/execution/executionResult';
import { CancellablePromise } from 'api/cancellablePromise';
import SubscriptionTracker from 'components/utils/SubscriptionTracker';
import { Compute, Connector, Namespace } from 'config/types';
import { ParsedSqlStatement } from 'parse/sqlStatementsParser';
import { getStatementsParser } from 'parse/utils';

type SessionPropertyValue = string | number | boolean | null | undefined;

export interface SessionProperty {
  defaultValue: SessionPropertyValue[];
  help_text?: string;
  key: string;
  multiple?: boolean;
  nice_name?: string;
  options?: SessionPropertyValue[];
  type?: string;
  value: SessionPropertyValue[];
}

export interface Session {
  configuration?: { [key: string]: string };
  http_addr?: string;
  id: number;
  properties: SessionProperty[];
  reuse_session?: boolean;
  session_id: string;
  type: string;
}

export interface ResultMeta {
  name: string;
  type: string;
  comment?: string | null;
}

export interface ResultApiResponse {
  data: (number | string)[][];
  has_more: boolean;
  isEscaped: boolean;
  meta: ResultMeta[];
  type: ResultType;
}

export interface ResultSizeApiResponse {
  rows: number | null;
  size: number | null;
  message?: string;
}

interface FetchResultData extends ExecutableContext {
  rows: number;
  startOver?: boolean;
}

interface FetchLogsData extends ExecutableContext {
  full_log: string;
  jobs?: string;
  from: number;
}

interface ExecuteData extends ExecutableContext {
  executable: string;
}

export interface ExecuteLogsApiResponse {
  logs: string;
  isFullLogs: boolean;
  jobs: ExecutionJob[];
}

export interface ExecutionJob {
  finished: boolean;
  name: string;
  percentJob?: number;
  started: boolean;
  url: string;
}

export interface ExecutionHandle {
  end?: { row: number; col: number };
  guid?: string;
  has_more_statements?: boolean;
  has_result_set?: boolean;
  log_context?: unknown;
  modified_row_count?: number;
  operation_type?: number;
  previous_statement_hash?: string;
  secret?: string;
  session_guid?: string;
  session_id?: number;
  session_type?: string;
  start?: { row: number; col: number };
  statement?: string;
  statement_id: number;
  statements_count?: number;
  sync?: boolean;
  result?: ResultApiResponse;
}

export interface ExecutionHistory {
  id: number;
  uuid?: string;
  parentUuid?: string;
}

export interface ExecuteApiResponse {
  handle: ExecutionHandle;
  history?: ExecutionHistory;
}

export interface ExecuteStatusApiResponse {
  result?: ResultApiResponse; // For streaming
  status: string;
  message?: string;
  has_result_set?: boolean;
}

export interface ExecuteApiOptions {
  executable: SqlExecutable;
  silenceErrors?: boolean;
}

export interface AuthRequest {
  auth: boolean;
  message?: string;
}

const CANCEL_STATEMENT_API = '/api/editor/cancel_statement';
const CHECK_STATUS_API = '/api/editor/check_status';
const CLOSE_SESSION_API = '/api/editor/close_session';
const CLOSE_STATEMENT_API = '/api/editor/close_statement';
const CREATE_SESSION_API = '/api/editor/create_session';
const EXECUTE_API_PREFIX = '/api/editor/execute/';
const FETCH_RESULT_DATA_API = '/api/editor/fetch_result_data';
const FETCH_RESULT_SIZE_API = '/api/editor/fetch_result_size';
const GET_LOGS_API = '/api/editor/get_logs';

export const createSession = async (options: {
  type: string;
  properties?: SessionProperty[];
  silenceErrors?: boolean;
}): Promise<AuthRequest | Session> => {
  const data = {
    session: JSON.stringify({ type: options.type, properties: options.properties || [] })
  };

  const responsePromise = post<
    {
      session?: Session;
      status: number;
      message?: string;
    },
    { session: string }
  >(CREATE_SESSION_API, data, {
    silenceErrors: !!options.silenceErrors,
    ignoreSuccessErrors: true
  });

  const response = await responsePromise;
  if (response.status === 401) {
    return { auth: true, message: response.message };
  }
  if (successResponseIsError(response)) {
    throw new Error(extractErrorMessage(response));
  }

  if (!response.session) {
    throw new Error('No session returned.');
  }
  return response.session;
};

export const closeSession = async (options: {
  session: Session;
  silenceErrors?: boolean;
}): Promise<void> => {
  const data = {
    session: JSON.stringify(options.session)
  };

  await post<void, { session: string }>(CLOSE_SESSION_API, data, {
    silenceErrors: !!options.silenceErrors
  });
};

export const executeStatement = async (options: ExecuteApiOptions): Promise<ExecuteApiResponse> => {
  const executable = options.executable;
  const url = EXECUTE_API_PREFIX + executable.executor.connector().dialect;

  const data = (await executable.toContext()) as ExecuteData;
  data.executable = executable.toJson();

  const executePromise = post<
    {
      handle: ExecutionHandle;
      history_id?: number;
      history_uuid?: string;
      history_parent_uuid?: string;
      result?: ResultApiResponse;
    },
    ExecuteData
  >(url, data, {
    silenceErrors: !!options.silenceErrors
  });

  executable.addCancellable({
    cancel: async () => {
      if (
        executable.status !== ExecutionStatus.running &&
        executable.status !== ExecutionStatus.streaming
      ) {
        // executable.status seems to have been set to 'canceling' so ignoring for now
        // return;
      }
      try {
        const response = await executePromise;
        if (options.executable.handle !== response.handle) {
          options.executable.handle = response.handle;
        }
        if (response.handle) {
          await cancelStatement(options);
        }
      } catch (err) {}
    }
  });

  const response = await executePromise;

  if (!response.handle) {
    throw new Error('No handle in execute response');
  }

  response.handle.result = response.result;

  const cleanedResponse: ExecuteApiResponse = {
    handle: response.handle
  };

  if (typeof response.history_id !== 'undefined') {
    cleanedResponse.history = {
      id: response.history_id,
      uuid: response.history_uuid,
      parentUuid: response.history_parent_uuid
    };
  }

  return cleanedResponse;
};

export const executeSingleStatement = ({
  database = 'default',
  connector,
  namespace,
  compute,
  statement
}: {
  connector: Connector;
  namespace: Namespace;
  compute: Compute;
  statement: string;
  database?: string;
}): CancellablePromise<{ meta?: ResultMeta[]; rows?: ResultRow[] }> => {
  const subTracker = new SubscriptionTracker();

  const promise = new CancellablePromise<{
    meta: ResultMeta[];
    rows: ResultRow[];
  }>((resolve, reject, onCancel) => {
    onCancel(() => reject('Cancelled'));

    const executor = new Executor({
      database: (() => database) as KnockoutObservable<string>,
      connector: (() => connector) as KnockoutObservable<Connector>,
      namespace: (() => namespace) as KnockoutObservable<Namespace>,
      compute: (() => compute) as KnockoutObservable<Compute>
    });

    let parsedStatement: ParsedSqlStatement | undefined = undefined;

    try {
      const parsedStatements = getStatementsParser(connector).parse(statement);
      parsedStatement = parsedStatements[0];
    } catch {}

    if (!parsedStatement) {
      reject('Failed parsing statement');
      return;
    }
    const sqlExecutable = new SqlExecutable({ executor, database, parsedStatement });
    executor.setExecutables([sqlExecutable]);

    subTracker.subscribe<ExecutableTransitionedEvent>(
      EXECUTABLE_TRANSITIONED_TOPIC,
      ({ executable, newStatus }) => {
        if (executable.id === sqlExecutable.id && newStatus === ExecutionStatus.failed) {
          reject('Execution failed.');
        } else if (
          executable.isSuccess() &&
          executable.handle &&
          !executable.handle.has_result_set
        ) {
          resolve();
        }
      }
    );

    subTracker.subscribe<ExecutableResultUpdatedEvent>(
      EXECUTABLE_RESULT_UPDATED_TOPIC,
      executionResult => {
        const { executable, rows, meta } = executionResult;
        if (executable.id === sqlExecutable.id && rows.length) {
          resolve({ rows, meta });
        }
      }
    );

    subTracker.addDisposable({
      dispose() {
        sqlExecutable.cancel().catch();
      }
    });

    sqlExecutable.execute().catch(reject);
  });

  promise.finally(() => {
    subTracker.dispose();
  });

  return promise;
};

export const cancelStatement = async (options: ExecuteApiOptions): Promise<void> => {
  const data = await options.executable.toContext();

  await post<void, ExecutableContext>(CANCEL_STATEMENT_API, data, {
    silenceErrors: !!options.silenceErrors
  });
};

export const closeStatement = async (options: ExecuteApiOptions): Promise<void> => {
  if (!options.executable.operationId) {
    return;
  }
  const data = { operationId: options.executable.operationId };
  await post<void, { operationId: string }>(CLOSE_STATEMENT_API, data, {
    silenceErrors: !!options.silenceErrors
  });
};

export const checkExecutionStatus = async (
  options: ExecuteApiOptions
): Promise<ExecuteStatusApiResponse> => {
  if (!options.executable.operationId) {
    throw new Error('No operationId given.');
  }

  const data = { operationId: options.executable.operationId };
  const responsePromise = post<
    {
      query_status?: ExecuteStatusApiResponse;
      status: number;
      message?: string;
    },
    { operationId: string }
  >(CHECK_STATUS_API, data, {
    silenceErrors: !!options.silenceErrors
  });

  options.executable.addCancellable(responsePromise);

  const response = await responsePromise;

  if (response.query_status) {
    return response.query_status;
  }

  if (response.status === -3) {
    return { status: ExecutionStatus.expired };
  }

  return { status: ExecutionStatus.failed, message: response.message };
};

export const fetchResults = async (options: {
  executable: SqlExecutable;
  rows: number;
  startOver?: boolean;
  silenceErrors?: boolean;
}): Promise<ResultApiResponse | undefined> => {
  const data = (await options.executable.toContext()) as FetchResultData;
  data.rows = options.rows;
  data.startOver = options.startOver;

  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  const transformResponse = (response: unknown) => JSON.bigdataParse(response);

  const response = await post<DefaultApiResponse & { result?: ResultApiResponse }>(
    FETCH_RESULT_DATA_API,
    data,
    {
      silenceErrors: !!options.silenceErrors,
      transformResponse
    }
  );

  return response.result;
};

export const fetchResultSize = async (
  options: ExecuteApiOptions
): Promise<ResultSizeApiResponse> => {
  const data = await options.executable.toContext();

  const responsePromise = post<
    {
      result: ResultSizeApiResponse;
    },
    ExecutableContext
  >(FETCH_RESULT_SIZE_API, data, {
    silenceErrors: !!options.silenceErrors
  });

  options.executable.addCancellable(responsePromise);
  const response = await responsePromise;
  return response.result;
};

export const fetchLogs = async (options: {
  executable: SqlExecutable;
  silenceErrors?: boolean;
  fullLog: string;
  jobs?: ExecutionJob[];
  from?: number;
}): Promise<ExecuteLogsApiResponse> => {
  const data = (await options.executable.toContext()) as FetchLogsData;
  data.full_log = options.fullLog;
  data.jobs = options.jobs && JSON.stringify(options.jobs);
  data.from = options.from || 0;

  const responsePromise = post<
    {
      logs: string;
      isFullLogs: boolean;
      jobs: ExecutionJob[];
      progress: number;
      status: number;
      message?: string;
    },
    FetchLogsData
  >(GET_LOGS_API, data, {
    silenceErrors: !!options.silenceErrors
  });

  options.executable.addCancellable(responsePromise);
  const response = await responsePromise;
  return {
    logs: (response.status === 1 && response.message) || response.logs || '',
    jobs: response.jobs || [],
    isFullLogs: response.isFullLogs
  };
};
