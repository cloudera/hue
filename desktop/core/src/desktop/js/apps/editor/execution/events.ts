// Licensed to Cloudera, Inc. under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  Cloudera, Inc. licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { Session } from 'apps/editor/execution/api';
import ExecutionResult from './executionResult';
import SqlExecutable, { ExecutionStatus } from './sqlExecutable';
import ExecutionLogs from './executionLogs';

export const EXECUTABLE_TRANSITIONED_TOPIC = 'hue.executable.status.transitioned';
export interface ExecutableTransitionedEvent {
  newStatus: ExecutionStatus;
  oldStatus: ExecutionStatus;
  executable: SqlExecutable;
}

export const EXECUTABLE_UPDATED_TOPIC = 'hue.executable.updated';
export type ExecutableUpdatedEvent = SqlExecutable;

export const EXECUTABLE_LOGS_UPDATED_TOPIC = 'hue.executable.logs.updated';
export type ExecutableLogsUpdatedEvent = ExecutionLogs;

export const EXECUTABLE_RESULT_UPDATED_TOPIC = 'hue.executable.result.updated';
export type ExecutableResultUpdatedEvent = ExecutionResult;

export const SHOW_SESSION_AUTH_MODAL_TOPIC = 'show.session.auth.modal';
export interface ShowSessionAuthModalEvent {
  message?: string;
  session: Session;
  resolve(): void;
  reject(): void;
}
