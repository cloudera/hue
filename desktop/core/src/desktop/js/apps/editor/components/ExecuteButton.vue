<!--
  Licensed to Cloudera, Inc. under one
  or more contributor license agreements.  See the NOTICE file
  distributed with this work for additional information
  regarding copyright ownership.  Cloudera, Inc. licenses this file
  to you under the Apache License, Version 2.0 (the
  "License"); you may not use this file except in compliance
  with the License.  You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
-->

<template>
  <HueButton
    v-if="loadingSession"
    key="loading-button"
    :small="true"
    :disabled="disabled"
    :title="I18n('Creating session')"
  >
    <i class="fa fa-fw fa-spinner fa-spin" /> {{ I18n('Loading') }}
  </HueButton>
  <HueButton
    v-if="showExecute"
    key="execute-button"
    :small="true"
    :primary="true"
    :disabled="disabled"
    @click="execute"
  >
    <i class="fa fa-play fa-fw" /> {{ I18n('Execute') }}
  </HueButton>

  <HueButton
    v-if="showStop && !stopping"
    key="stop-button"
    :small="true"
    :alert="true"
    @click="stop"
  >
    <i class="fa fa-stop fa-fw" />
    <span v-if="waiting">{{ I18n('Stop batch') }}</span>
    <span v-else>{{ I18n('Stop') }}</span>
  </HueButton>

  <HueButton v-if="showStop && stopping" key="stopping-button" :small="true" :alert="true">
    <i class="fa fa-fw fa-spinner fa-spin" /> {{ I18n('Stopping') }}
  </HueButton>
</template>

<script lang="ts">
  import { defineComponent, PropType, ref, toRefs, watch } from 'vue';

  import { EXECUTE_ACTIVE_EXECUTABLE_TOPIC, ExecuteActiveExecutableEvent } from './events';
  import { Session } from 'apps/editor/execution/api';
  import {
    EXECUTABLE_TRANSITIONED_TOPIC,
    EXECUTABLE_UPDATED_TOPIC,
    ExecutableTransitionedEvent,
    ExecutableUpdatedEvent
  } from 'apps/editor/execution/events';
  import sessionManager from 'apps/editor/execution/sessionManager';
  import SqlExecutable, { ExecutionStatus } from 'apps/editor/execution/sqlExecutable';
  import HueButton from 'components/HueButton.vue';
  import SubscriptionTracker from 'components/utils/SubscriptionTracker';
  import huePubSub from 'utils/huePubSub';
  import I18n from 'utils/i18n';

  const WHITE_SPACE_REGEX = /^\s*$/;

  export default defineComponent({
    name: 'ExecuteButton',
    components: {
      HueButton
    },
    props: {
      executable: {
        type: Object as PropType<SqlExecutable | undefined>,
        default: undefined
      },
      beforeExecute: {
        type: Function as PropType<((executable: SqlExecutable) => Promise<void>) | undefined>,
        default: undefined
      }
    },
    emits: [
      'execute-failed',
      'execute-started',
      'execute-successful',
      'executable-updated',
      'execute-stopping'
    ],
    setup(props, { emit }) {
      const { executable, beforeExecute } = toRefs(props);
      const subTracker = new SubscriptionTracker();

      let lastSession = null as Session | null;

      const stopping = ref(false);
      const loadingSession = ref(true);
      const partOfRunningExecution = ref(false);
      const status = ref<ExecutionStatus>(ExecutionStatus.ready);
      const hasStatement = ref(false);

      const execute = async (): Promise<void> => {
        huePubSub.publish('hue.ace.autocompleter.hide');
        if (!executable.value) {
          return;
        }
        if (beforeExecute.value) {
          await beforeExecute.value(executable.value as SqlExecutable);
        }
        await executable.value.reset();
        emit('execute-started', executable.value);
        executable.value.execute();
      };

      const stop = async (): Promise<void> => {
        if (stopping.value || !executable.value) {
          return;
        }
        emit('execute-stopping', executable.value);
        stopping.value = true;
        await executable.value.cancelBatchChain(true);
        stopping.value = false;
      };

      const updateFromExecutable = (executable: SqlExecutable): void => {
        const waitForSession =
          !lastSession || lastSession.type !== executable.executor.connector().type;
        status.value = executable.status;
        hasStatement.value =
          !executable.parsedStatement ||
          !WHITE_SPACE_REGEX.test(executable.parsedStatement.statement);
        partOfRunningExecution.value = executable.isPartOfRunningExecution();
        if (waitForSession) {
          loadingSession.value = true;
          lastSession = null;
          sessionManager.getSession({ type: executable.executor.connector().id }).then(session => {
            lastSession = session;
            loadingSession.value = false;
          });
        }
      };

      watch(
        executable,
        newVal => {
          if (newVal) {
            updateFromExecutable(newVal as SqlExecutable);
          }
        },
        { immediate: true }
      );

      subTracker.subscribe<ExecutableUpdatedEvent>(EXECUTABLE_UPDATED_TOPIC, updatedExecutable => {
        const active = executable.value && executable.value.id === updatedExecutable.id;
        if (active) {
          updateFromExecutable(updatedExecutable);
        }
        emit('executable-updated', { executable: updatedExecutable, active });
      });

      subTracker.subscribe<ExecutableTransitionedEvent>(EXECUTABLE_TRANSITIONED_TOPIC, event => {
        if (event.executable.id === executable.value?.id) {
          if (
            event.newStatus === ExecutionStatus.available ||
            event.newStatus === ExecutionStatus.streaming ||
            event.newStatus === ExecutionStatus.success
          ) {
            emit('execute-successful', executable.value);
          } else if (event.newStatus === ExecutionStatus.failed) {
            emit('execute-failed', executable.value);
          }
        }
      });

      subTracker.subscribe<ExecuteActiveExecutableEvent>(
        EXECUTE_ACTIVE_EXECUTABLE_TOPIC,
        async eventExecutable => {
          if (executable.value && executable.value.id === eventExecutable.id) {
            await execute();
          }
        }
      );

      return {
        execute,
        subTracker,
        stop,
        stopping,
        loadingSession,
        partOfRunningExecution,
        status,
        hasStatement,
        I18n
      };
    },
    computed: {
      waiting(): boolean {
        return !!(this.executable && this.executable.isReady() && this.partOfRunningExecution);
      },
      disabled(): boolean {
        return this.loadingSession || !this.executable || !this.hasStatement;
      },
      showExecute(): boolean {
        return (
          !!this.executable &&
          !this.waiting &&
          !this.loadingSession &&
          this.status !== ExecutionStatus.running &&
          this.status !== ExecutionStatus.streaming
        );
      },
      showStop(): boolean {
        return (
          this.status === ExecutionStatus.running ||
          this.status === ExecutionStatus.streaming ||
          this.waiting
        );
      }
    }
  });
</script>
