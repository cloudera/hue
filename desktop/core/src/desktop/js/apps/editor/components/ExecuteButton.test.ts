// Licensed to Cloudera, Inc. under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  Cloudera, Inc. licenses this file
// to you under the Apache License, Version 2.0 (the
// 'License'); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an 'AS IS' BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { nextTick } from 'vue';
import { mount, shallowMount } from '@vue/test-utils';

import { Session } from '../execution/api';
import { EXECUTABLE_UPDATED_TOPIC, ExecutableUpdatedEvent } from 'apps/editor/execution/events';
import huePubSub from 'utils/huePubSub';
import SqlExecutable, { ExecutionStatus } from 'apps/editor/execution/sqlExecutable';
import sessionManager from 'apps/editor/execution/sessionManager';
import ExecuteButton from './ExecuteButton.vue';
import noop from 'utils/timing/noop';

describe('ExecuteButton.vue', () => {
  it('should render', () => {
    const wrapper = shallowMount(ExecuteButton);
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should show execute once the session is loaded', async () => {
    const spy = jest
      .spyOn(sessionManager, 'getSession')
      .mockReturnValue(Promise.resolve({ type: 'foo' } as Session));

    const mockExecutable = {
      cancel: noop,
      cancelBatchChain: noop,
      execute: noop,
      isPartOfRunningExecution: () => false,
      isReady: () => true,
      reset: noop,
      nextExecutable: {},
      executor: {
        defaultLimit: noop,
        connector: () => ({
          id: 'foo'
        })
      },
      status: ExecutionStatus.ready
    } as SqlExecutable;

    const wrapper = shallowMount(ExecuteButton, {
      props: {
        executable: mockExecutable
      }
    });

    await nextTick();

    expect(spy).toHaveBeenCalled();
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should handle execute and stop clicks', async () => {
    const spy = jest
      .spyOn(sessionManager, 'getSession')
      .mockReturnValue(Promise.resolve({ type: 'foo' } as Session));
    let executeCalled = false;
    let cancelCalled = false;
    const mockExecutable = {
      cancel: () => {
        cancelCalled = true;
      },
      cancelBatchChain: () => {
        cancelCalled = true;
      },
      execute: async () => {
        executeCalled = true;
      },
      isPartOfRunningExecution: () => false,
      isReady: () => true,
      reset: noop,
      nextExecutable: {},
      executor: {
        defaultLimit: noop,
        connector: () => ({
          id: 'foo',
          type: 'foo'
        })
      },
      status: ExecutionStatus.ready
    } as SqlExecutable;

    const wrapper = mount(ExecuteButton, {
      props: {
        executable: mockExecutable
      }
    });

    await nextTick();

    expect(spy).toHaveBeenCalled();

    // Click play
    expect(executeCalled).toBeFalsy();
    expect(wrapper.get('button').text()).toContain('Execute');
    wrapper.get('button').trigger('click');

    await nextTick();

    expect(executeCalled).toBeTruthy();
    mockExecutable.status = ExecutionStatus.running;
    huePubSub.publish<ExecutableUpdatedEvent>(
      EXECUTABLE_UPDATED_TOPIC,
      mockExecutable as SqlExecutable
    );

    await nextTick();

    // Click stop
    expect(cancelCalled).toBeFalsy();
    expect(wrapper.get('button').text()).toContain('Stop');
    wrapper.get('button').trigger('click');

    await nextTick();

    expect(cancelCalled).toBeTruthy();
  });
});
