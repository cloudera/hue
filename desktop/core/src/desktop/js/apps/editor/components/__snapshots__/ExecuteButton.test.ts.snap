// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExecuteButton.vue should render 1`] = `
<div
  data-v-app=""
>
  
  <hue-button-stub
    alert="false"
    borderless="false"
    disabled="true"
    large="false"
    primary="false"
    small="true"
    title="Creating session"
  />
  <!--v-if-->
  <!--v-if-->
  <!--v-if-->
  
</div>
`;

exports[`ExecuteButton.vue should show execute once the session is loaded 1`] = `
<div
  data-v-app=""
>
  
  <!--v-if-->
  <hue-button-stub
    alert="false"
    borderless="false"
    disabled="false"
    large="false"
    primary="true"
    small="true"
  />
  <!--v-if-->
  <!--v-if-->
  
</div>
`;
